granulate_utils/glogger/__init__.py

```diff
@@ -0,0 +1,189 @@
+        self.stop_event = threading.Event()
+        self.time_fn = time.time
+        self.server_address = server_address
+        self.session = requests.Session()
+        self.last_flush_time = 0.0
+        self.flush_thread = threading.Thread(target=self._flush_loop, daemon=True, name="Logs flusher")
+        self.start()
+
+    def start(self) -> None:
+        """Called to start the flush thread."""
+        self.flush_thread.start()
+
+    def emit(self, record: LogRecord) -> None:
+        # Called while lock is acquired
+        item = self.dump_record_locked(record)
+        self.messages_buffer.append(item)
+
+    def dump_record_locked(self, record: LogRecord) -> str:
+        # `format` is required to fill in some record members such as `message` and `asctime`.
+        self.format(record)
+        record_dict = self.dictify_record(record)
+        if record.truncated:  # type: ignore
+            record_dict["truncated"] = True
+        record_dict["serial_no"] = self.messages_buffer.next_serial_no
+        return json.dumps(record_dict)
+
+    def format(self, record: LogRecord) -> str:
+        """Override to customize record fields."""
+        s = super().format(record)
+        if len(record.message) > self.max_message_size:
+            record.message = record.message[: self.max_message_size]
+            record.truncated = True  # type: ignore
+        else:
+            record.truncated = False  # type: ignore
+        return s
+
+    def dictify_record(self, record):
+        formatted_timestamp = datetime.utcfromtimestamp(record.created).isoformat()
+        return {
+            "severity": record.levelno,
+            "timestamp": formatted_timestamp,
+            "logger_name": record.name,
+            "message": record.message,
+        }
+
+    def _flush_loop(self) -> None:
+        self.last_flush_time = self.time_fn()
+        while not self.stop_event.is_set():
+            if self.should_flush():
+                self.flush()
+            self.stop_event.wait(1.0)
+
+    @property
+    def time_since_last_flush(self) -> float:
+        return self.time_fn() - self.last_flush_time
+
+    def should_flush(self) -> bool:
+        return (
+            self.messages_buffer.count > 0
+            and (self.messages_buffer.utilized >= self.flush_threshold)
+            or (self.time_since_last_flush >= self.flush_interval)
+        )
+
+    def flush(self) -> None:
+        # Allow configuring max_tries via class member. Alternatively we could decorate _send_logs in __init__ but that
+        # would be easier to miss and less flexible.
+        send = backoff.on_exception(backoff.expo, exception=RequestException, max_tries=self.max_send_tries)(
+            self._send_logs
+        )
+
+        self.last_flush_time = self.time_fn()
+        try:
+            batch, response = send()
+            response.raise_for_status()
+        except Exception:
+            logger.exception("Error posting to server")
```

Comment:
This really scares me that it can cause issues in multithreading - WDYT?

Reply:
- This is called when lock is not held, so buffer is not being modified, so corruption won't happen.
- The lock is RLock (ie reentrant) so if it were held it won't cause a deadlock.

Deadlock does not worry me. The bigger fear is corruption which in this case won't happen. But I see now that it can definitely happen with the "dropped" message in the MessagesBuffer class...